runtime/compiler/optimizer/JProfilingBlock.cpp

```diff
@@ -695,9 +695,9 @@ int32_t TR_JProfilingBlock::processCFGForCounting(BlockParents &parent, TR::Bloc
             }
          // if the source of the edge has a single source we can count the source
          else if (block != to && !block->isOSRCatchBlock() &&
-             (block->getSuccessors().size() == 1 && block->getExceptionSuccessors().size() == 0)
+             ((block->getSuccessors().size() == 1 && block->getExceptionSuccessors().size() == 0)
              || (block->getSuccessors().size() == 0 && block->getExceptionSuccessors().size() == 1)
-             || block->isOSRInduceBlock() || block->isOSRCodeBlock())
+             || block->isOSRInduceBlock() || block->isOSRCodeBlock()))
             {
```

Comment:
Could I get someone familiar with this code to review this? This is a change in behaviour, but reading the code I think this is what we really wanted to check for? @r30shah/@jdmpapin  perhaps?

Reply:
@fjeremic Yes, I think the change you made should be the correct behaviour. Seems like current one could incorrectly put the counter in same block in cases of self calling edge where we should really consider it a critical edge and insert the block to add counter. However, we would not have satisfied successor/predecessor condition. I am not sure about OSR CodeBlock/InduceBlock.

Reply:
I agree with @r30shah that the `block != to` condition is trying to exclude self-loops from this case (to force the self-loop edge to be split for counting), and that before your changes the full `else if` condition will sometimes fail to impose the `block != to` requirement. It will also sometimes fail to reject `block` when it's an OSR catch block.

So it would seem that before your change, we would unintentionally enter here in the following cases:
1. OSR induce or code block that is also an OSR catch block
2. OSR induce or code block with a self-loop
3. Exception handler whose only successor is a self-loop exception edge

I think 1 and 2 are impossible, and 3 is either impossible or very hard to generate.

I did some poking and it looks like the OSR induce block is listed here because it has a "regular" outgoing edge to the method exit block, but that edge is never followed. Instead it always follows the exception edge to the OSR catch block. I'm not sure why OSR code blocks are listed as well - it looks like those should always have a single successor anyway (either the caller's OSR code block, or the method exit).

As an aside, I'm not sure that this case (or the next) needs to exclude self-loops at all (though this isn't necessarily a suggestion to allow them). A block that is its own only successor will traverse its self-loop exactly once each time (just after) it runs (barring a thrown exception without any visible handlers, but any single-successor block has the same caveat). A block that is its own only predecessor is unreachable, so it should never even occur in the CFG.

Reply:
Leaving this review unresolved for purposes of code archeology so if someone ever stumbles upon this change they will see the above excellent explanation. Thanks!